libs/entitlement/src/main/java/org/elasticsearch/entitlement/runtime/policy/entitlements/FilesEntitlement.java

```diff
@@ -29,8 +34,104 @@ public enum Mode {
         READ_WRITE
     }
 
-    public record FileData(String path, Mode mode) {
+    public enum BaseDir {
+        CONFIG,
+        DATA
+    }
+
+    public sealed interface FileData {
+
+        final class AbsolutePathFileData implements FileData {
```

Comment:
The implementations could still be records? Then you won't have to implement equals/hashcode.

Reply:
I've quite torn about that, because with these as records they could be used directly and it will expose the members (I really like strict encapsulation)

Reply:
They can still be package private records, they don't need to be public, so it is no different than now with the accessors.